File: src/main/java/com/github/mgramin/sqlboot/model/uri/impl/DbUri.java
/*
 * The MIT License (MIT)
 * <p>
 * Copyright (c) 2016-2019 Maksim Gramin
 * <p>
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * <p>
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.github.mgramin.sqlboot.model.uri.impl;

import static java.util.Arrays.asList;

import com.github.mgramin.sqlboot.exceptions.BootException;
import com.github.mgramin.sqlboot.model.uri.Uri;
import java.net.URI;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by maksim on 12.06.16.
 */
public final class DbUri implements Uri {

    private final String type;
    private final String action;
    private final List<String> objects;
    private final Boolean recursive;
    private final Map<String, String> params = new LinkedHashMap<>();

    public DbUri(final String type, String ... objects) {
        this.type = type;
        this.objects = Arrays.asList(objects);
        this.action = null;
        this.recursive = false;
    }

    public DbUri(final String type, final List<String> objects) {
        this.type = type;
        this.objects = objects;
        this.action = null;
        this.recursive = false;
    }

    public DbUri(final String uriString) throws BootException {
        try {
            URI uri = new URI(uriString);
            String pathString = uri.getPath().replace("*", "%");
            String queryString = uri.getQuery();

            List<String> list = asList(pathString.split("[/]"));
            type = list.get(0);
            if (list.size() == 1) {
                objects = asList("%");
            } else {
                objects = asList(list.get(1).split("[.]"));
            }
            if (list.size() == 3) {
                action = list.get(2);
            } else {
                action = null;
            }
            recursive = pathString.charAt(pathString.length() - 1) == '/';

            if (queryString != null)
                for (String s : queryString.split("&")) {
                    params.put(s.split("=")[0], s.split("=")[1]);
            }
        } catch (Exception e) {
            throw new BootException(e);
        }
    }

    @Override
    public String type() {
        return type;
    }

    @Override
    public List<String> path() {
        return objects;
    }

    @Override
    public String path(Integer index) {
        if (index > objects.size()-1) {
            return "%";
        } else {
            return objects.get(index);
        }
    }

    @Override
    public Boolean recursive() {
        return recursive;
    }

    @Override
    public Map<String, String> params() {
        return params;
    }

    public String action() {
        return this.action;
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder(type + "/" + String.join(".", objects));
        if (action != null && !action.equals("create"))
            result.append("/").append(action);
        if (recursive != null && recursive)
            result.append("/");
        if (!params.isEmpty()) {
            result.append("?");
            int i = 0;
            for (Map.Entry<String, String> stringStringEntry : params.entrySet()) {
                if (++i!=1)
                    result.append("&");
                result.append(stringStringEntry.getKey()).append("=").append(stringStringEntry.getValue());
            }
        }
        return result.toString().replace("%", "*");
    }

}

File: src/main/kotlin/com/github/mgramin/sqlboot/model/uri/impl/DbUri.kt
/*
 * BSD 3-Clause License
 *
 * Copyright (c) 2019, CROC Inc.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package com.github.mgramin.sqlboot.model.uri.impl

import com.github.mgramin.sqlboot.exceptions.BootException
import com.github.mgramin.sqlboot.model.uri.Uri
import org.springframework.web.util.UriComponentsBuilder


/**
 * Created by maksim on 12.06.16.
 */
class DbUri : Uri {

    private val connection: String
    private val type: String
    // TODO rename to PATH
    private val objects: List<String>
    private val params: Map<String, String>
    private val action: String

    constructor(connection: String, type: String, objects: List<String>) {
        this.connection = connection
        this.type = type
        this.objects = objects
        this.action = ""
        this.params = emptyMap()
    }

    constructor(uriString: String) {
        try {
            val uri = UriComponentsBuilder.fromUriString(uriString).build()
            val pathMap: Map<Int, String> = uri.path
                    .split("/")
                    .filter { it.isNotEmpty() }
                    .mapIndexed { index, element -> index to element }
                    .toMap()

            connection = pathMap.getValue(0)
            type = pathMap.getValue(1)

            objects = if (pathMap.containsKey(2)) {
                pathMap[2]!!.split(".")
            } else {
                emptyList()
            }

            action = pathMap.getOrDefault(3, "")

            params = if (uri.query != null)
                uri.queryParams
                        .map { it.key to it.value.first() }
                        .toMap()
            else emptyMap()

        } catch (e: Exception) {
            throw BootException(e)
        }
    }

    override fun type() = type

    override fun path() = objects

    override fun path(index: Int) =
            if (index > objects.size - 1) {
                "%"
            } else {
                objects[index]
            }

    override fun params() = params

    override fun action() = this.action

    override fun connection() = this.connection

    override fun toString(): String {
        val result = StringBuilder(connection + "/" + type + "/" + objects.joinToString("."))
        if (action != "" && action != "create")
            result.append("/").append(action)
        if (!params.isEmpty()) {
            result.append("?")
            var i = 0
            for ((key, value) in params) {
                if (++i != 1)
                    result.append("&")
                result.append(key).append("=").append(value)
            }
        }
        return result.toString().replace("%", "*")
    }

}